### PR TITLE
Eisner ensemble

### PR DESCRIPTION
**Pull request by MihaiSurdeanu**

### CODE REVIEW

**Comment by MihaiSurdeanu** [comment]
@kwalcock : this is ready for your review. Thanks!

**Comment by MihaiSurdeanu** [comment]
Also, when was 8.4.8 released? Some changes in CHANGES.md may belong in the next 8.4.9 release.

**Comment by kwalcock** [comment]
8.4.8 is from 2/15/2022 @ 12:32 PM.  Not sure what time zone.

**Comment by MihaiSurdeanu** [comment]
Thanks! I'll adjust CHANGES.

**Comment by MihaiSurdeanu** [comment]
Also, the previous PR probably deserves your attention...

**Review by kwalcock — APPROVED** [review]
I don't think anything especially needs to change.  There were a couple places, especially with conversion of strings and encoding of those heads and deps where I imagined the potential for error, though, and maybe they can be double checked.  The rest are probably more notes for myself and should have been kept that way.

**Comment by kwalcock** [comment]
> Also, the previous PR probably deserves your attention...

#603 seemed reasonable.

File: main/src/main/scala/org/clulab/dynet/Eisner.scala [hunk 1]
```diff
@@ -262,16 +218,157 @@ class Eisner {
...
+  def generateOutput(top: Option[Span],
+                     scores: IndexedSeq[IndexedSeq[(String, Float)]],
+                     dependencies: Array[Array[Dependency]],
+                     generateRelativeHeads: Boolean): IndexedSeq[(Int, String)] = {
+
+    val heads = new Array[(Int, String)](scores.size)
+    if(top.nonEmpty) {
+      // Eisner correctly produced a full tree
+      for(dep <- top.get.dependencies) {
+        val head =
```
[the remark this answers is not in the document]
Comment (MihaiSurdeanu) [reply main/src/main/scala/org/clulab/dynet/Eisner.scala:233]: In this case, we are screwed anyway... So, I'd say no :)

File: main/src/main/scala/org/clulab/dynet/Eisner.scala [hunk 2]
```diff
@@ -262,16 +218,157 @@ class Eisner {
...
+            // we are storing absolute heads, starting at offset 0
+            dep.head - 1
+          }
+        val label = dep.label
+        heads(dep.mod - 1) = Tuple2(head, label)
+      }
+    } else {
+      // Eisner failed to produce a complete tree; revert to the greedy inference
+      for(i <- scores.indices) {
+        val relativeHead = scores(i).minBy(- _._2)._1.toInt
```
[the remark this answers is not in the document]
Comment (MihaiSurdeanu) [reply main/src/main/scala/org/clulab/dynet/Eisner.scala:247]: I think these can be replaced by maxBy!

**Comment by kwalcock** [comment]
If it's OK, I'll merge and then implement in a separate round.

**Pull request merged**